man/man8/sendmail.8
.TH SENDMAIL 8 "23 November 98"
.SH NAME
MMDF \- sendmail replacement for ``real-sendmail''
.SH SYNOPSIS
/usr/lib/sendmail 
.RI "[ " flags " ] [ " address " "  ... " ]"
.br
.SH DESCRIPTION
.I sendmail
is a send-only sendmail emulator for mail-clients.
It provides the functionality required for humans and programs to send
mail via the standard or /usr/bin/mail user agents.
.PP
It accepts a mail stream on standard input with recipients specified on the
command line and synchronously forwards the message to MMDF system.


.SH OPTIONS
Most sendmail options are irrelevent to MMDF sendmail. Those marked ``ignored'' or
``default'' have no effect on mail transfer.  Those marked ``unsupported''
are fatal errors.  Those marked ``simulated'' are not errors, but the result
is for the program to exit with an informative message. A sort of fatal
non-error.
.br
All sendmail options not listed here are ``unsupported''.

.TP
.B \-ba
Go into ARPANET mode. All input lines must end with a CR-LF, and
all messages will be generated with a CR-LF at the end. Also, the ``From:''
and ``Sender:'' fields are examined for the name of the sender.

.TP
.B \-bs
Read SMTP commands from stdin.

.TP
.B \-bm
(default) Deliver mail in the usual way.

.TP
.B \-bv
(unsupported) Verify names only.

.TP
.B \-b\fP\fI*\fP
(unsuppoerted) \fP\fI*\fP anything else then above

.TP
.B \-f\fP\fIname\fP
Sets the name of the ``from'' person.

.TP
.B \-r
Same as -f.

.TP
.B \-F\fP\fIfullname\fP
Set the full name of the sender.

.TP
.B \-h\fP\fI\fN\fP
(ignored) Set the hop count to N.

.TP
.B \-t
Read  message, searching for recipients. ``To:'', `Cc:'', and ``Bcc:'' lines
will be scanned for people to send to. Any addresses  in  the  argument list
will be suppressed.

.TP
.B \-v
Go into verbose mode. 

.TP
.B \-T
(ignored)

.TP
.B \-C\fP\fIfile\fP 
(ignored) Use alternate configuration file.

.TP
.B \-c
(ignored)

.TP
.B \-d\fP\fIX\fP
(ignored) Set debugging value to X.

.TP
.B \-e
(ignored)

.TP
.B \-i
(ignored) Ignore dots alone on lines by themselves in incoming
messages. This should be set if you are reading data from a file.

.TP
.B \-m
(ignored) Ignore originator in alias.

.TP
.B \-n
(ignored) Do not do aliasing.

.TP
.B \-o\fP\fIx\fP \fP\fIvalue\fP
(ignored) Set option \fIx\fP to the specified \fIvalue\fP. This form
uses single character names only.

.TP
.B \-s
(ignored)

.TP
.B \-B\fP\fItype\fP
(ignored) Set the body type to \fItype\fP. Current legal values 7BIT or 8BITMIME.

.TP
.B \-N \fP\fIdsn\fP
(ignored) Set delivery status notification conditions to \fIdsn\fP, which can
be `never' for no notifications or a comma separated list of the
values `failure' to be notified if delivery failed, `delay' to be
notified if delivery is delayed, and `success' to be notified when the
message is successfully delivered. 

.TP
.B \-p\fP\fIprotocol\fP
(ignored) Set the name of the protocol used to receive the
message. This can be a simple protocol name such as ``UUCP'' or a
protocol and hostname, such as ``UUCP:ucbvax''.

.TP
.B \-U
(ignored) Initial (user) submission.  This should \fIalways\fP be
set when called from a user agent such as \fBMail\fP or \fBexmh\fP and
\fInever\fP be set when called by a network delivery agent such as
\fBrmail\fP.

.TP
.B \-V \fP\fIenvid\fP
(ignored) Set the original envelope id. This is propogated across
SMTP to servers that support DSNs and is returned in DSN-compliant
error messages.

.SH FILES
/usr/lib/sendmail
.br
@SBINDIR@/sendmail
.br

.SH "SEE ALSO"
RFC 821, RFC 822.


Changes
Changes beta 3
	* Small bug in ap_lex.c hip@a.cs.okstate.edu
	* jromine@ics.uci.edu some changes for Solaris 2.5, Sun Pro
	  4.2 Compiler
	* linux 2.2.7, glibc2.1, egcs 2.91 (redhat 6.0)
	* sco (first attempt)
	* jromine@ics.uci.edu, boundary fix smtpsrvr
	* Small bug in submit/lnk_getaddr.c krueger@mathematik.uni-kl.de

Wed Dec 31 13:14:34 MET 1997	Kai Krueger <krueger@mathematik.uni-kl.de>

	* create file h/config.h.in and moved every define to it.

	* DOASSIGN renamed to HAVE_ASSIGN. Autodetection by configure

	* SECURETTY renamed to HAVE_SECURETTY, Autodetection by configure

	* NODIAL renamed to HAVE_DIAL. default is 'no'
	  may be enanble by configure --enable-dial
	  if compiling phone/smphone channel it wil be turned on by configure.

	* NAMESERVER renamed to HAVE_NAMESERVER. Autodetection by configure.
	  May be overwriten by --disable-nameserver
          - tb_ns.4.2.c and tb_ns.fake.c merged into one file tb_ns.c and
	  seperated by #ifdef HAVE_NAMESERVER...
	  - removed TB_NS in Makefile.com	  

	* HAVE_TCP_WRAPPER renamed to HAVE_LIBWRAP. Autodetection by configure.
	
	* <time.h> included in config.h. It's location is determined by
	  configure

	* determination of <dir.h>, <dirent.h>, a.s.o is done by configure
	  definition an include has moved to config.h

	* determination of index/rindex done by configure and moved to config.h

	* definition of sigtype renamed to RETSIGTYPE. Detection done by
	  configure.

	* return-type of sprintf determined by configure. Define BSD_SPRINTF
	  if char * is returned.

	* uip/Makefile, src/Makefile, lib/Makefile Makefile.com and Makefile
	  are generated by configure.

	* Value of CH_TB in Makefile.com set by configure. If you want
	  sequential tables run configure with the argument --disable-dbm

	* Added define HAVE_SGTTY_H for the existens of <sgtty.h>

	* Added check for <varargs.h> as define HAVE_VARARGS_H.
	  Added check for variable number of arguments if <varargs.h> don't
	  exists. If Compiler don't supports variable number of arguments
	  NO_VARARGS is defined. (NO_VARARGS should be renamed)
	  configure ist checking for both of it. But if <varagrs.h> isn't
	  found lib/util/multcat.c and lib/util/multcpy.c could't be compiled.
	  
	* test configure on machine that dosn't support multiple arguments.
	  (check of NO_VARARGS)

	* test for dup2() and fcntl(fd1, F_DUPFD, fd2) done.
	  NODUP2 replaced by HAVE_DUP2 and
	  NOFCTNL replaced by HAVE_FCNTL_F_DUPFD


uip/other/sendmail.c
/*
**  A Sendmail fake.
*/

#include <signal.h>
#include <pwd.h>
#include "util.h"
#include "mmdf.h"
#include "cnvtdate.h"
#include "mm_io.h"

#if !defined(__STDC__) || defined(DECLARE_GETPWUID)
extern struct passwd *getpwuid ();
#endif /* DECLARE_GETPWUID */
#ifdef DECLARE_EXIT
extern int exit();
#endif /* DECLARE_EXIT */

extern	char	*locname, *locdomain;
extern	char	*chndfldir;

extern char *getmailid ();
extern char *getenv();
extern char *dupfpath();
extern char *multcat();

char	*SMTPSRVR = "smtpsrvr";

char	*FullName;	/* sender's full name */
char	*from;		/* sender's mail address */
char	subflags[256];	/* flags for submit */
int	watch;
int	verify;
int	extract;
int	badaddrs;
int	rewritefrom;

RETSIGTYPE	die();

/*ARGSUSED*/
main(argc, argv)
int argc;
char **argv;
{
	struct passwd  *pwdptr;
	register char *p;
	register char **av;
	struct rp_bufstruct thereply;
	int	retval;
	int     len;
    char real_user[128];
    char buf[256];	/* flags for submit */

	mmdf_init(argv[0], 0);

	if ((pwdptr = getpwuid (getuid())) == (struct passwd *) NULL)
		syserr("Unable to locate user's name");

	if ((from = getmailid(pwdptr -> pw_name)) == NULL)
		syserr("Unable to locate user's mailid");

	from = multcat(from, "@", locname, (char *)0);

	if (signal(SIGINT, SIG_IGN) != SIG_IGN)
		(void) signal(SIGINT, die);
	if (signal(SIGHUP, SIG_IGN) != SIG_IGN)
		(void) signal(SIGHUP, die);
	(void) signal(SIGTERM, die);
	(void) signal(SIGPIPE, die);

	FullName = getenv("NAME");

	av = argv;
	while ((p = *++av) != NULL && p[0] == '-') {
		switch (p[1]) {
		case 'b':	/* operations mode */
			switch (p[2]) {
			case 'a':	/* smtp on stdin */
			case 's':	/* smtp on stdin */
				smtp();
				exit(98);	/* should never happen */
			case 'm':	/* just send mail */
				continue;
			case 'v':	/* verify mode */
				verify++;
				continue;
			default:
				syserr("Invalid operation mode %c", p[2]);
			}
			continue;

		case 'f':	/* from address */
		case 'r':	/* obsolete -f flag */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || *p == '-'))
			{
				p = *++av;
				if (p == NULL || *p == '-') {
					syserr("No \"from\" person");
					av--;
					continue;
				}
			}
			if (rewritefrom) {
				syserr("More than one \"from\" person");
				continue;
			}
			from = p;
			rewritefrom++;
			continue;

		case 'F':	/* set full name */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || *p == '-'))
			{
				syserr("Bad -F flag");
				av--;
				continue;
			}
			FullName = p;
			continue;

		case 'h':	/* hop count */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || !isdigit(*p)))
			{
				syserr("Bad hop count (%s)", p);
				av--;
				continue;
			}
			continue;	/* Ignore */

		case 't':	/* read recipients from message */
			extract++;
			continue;


		case 'v':	/* give blow-by-blow description */
			watch++;
			continue;

        case 'B': /* body type: 7BIT or 8BITMIME */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || *p == '-'))
			{
				syserr("Bad -B flag");
				av--;
				continue;
			}
#ifdef HAVE_ESMTP
#endif  /* HAVE_ESMTP */
			continue;

        case 'N': /* delivery status notifications */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || *p == '-'))
			{
				syserr("Bad -N flag");
				av--;
				continue;
			}
#ifdef HAVE_ESMTP
#endif  /* HAVE_ESMTP */
              continue;
              
        case 'R': /* DSN RET: what to return */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || *p == '-'))
			{
				syserr("Bad -R flag");
				av--;
				continue;
			}
#ifdef HAVE_ESMTP
#endif  /* HAVE_ESMTP */
              continue;
              
        case 'V': /* DSN ENVID: set "original" envelope id */
			p += 2;
			if (*p == '\0' && ((p = *++av) == NULL || *p == '-'))
			{
				syserr("Bad -V flag");
				av--;
				continue;
			}
#ifdef HAVE_ESMTP
#endif  /* HAVE_ESMTP */
              continue;
              
        case 'T':	/* set timeout interval */
		case 'C':	/* select configuration file (already done) */
		case 'c':	/* connect to non-local mailers */
		case 'd':	/* debug */
		case 'e':	/* error message disposition */
		case 'i':	/* don't let dot stop me */
		case 'm':	/* send to me too */
		case 'n':	/* don't alias */
		case 'o':	/* set option, handled in line */
		case 's':	/* save From lines in headers */
			continue;
		}
	}

	setuid(getuid());

	if (verify && extract)
		syserr("Verify mode not supported on header components");

        memset(subflags, 0, sizeof(subflags));
        memset(buf, 0, sizeof(buf));
	strncpy(subflags, "ml", sizeof(subflags));
	if (rewritefrom)
		strcat(subflags, "t");
	if (watch)
		strcat(subflags, "w");
	if (!extract)
		strcat(subflags, "v");
	else
		strcat(subflags, "xto,cc,bcc*");
        snprintf(buf, sizeof(buf), "%sF%s*", subflags, pwdptr->pw_name );
 	strncpy(subflags, buf, sizeof(subflags));

	if (rp_isbad (mm_init ()) || rp_isbad (mm_sbinit ()))
		syserr("Unable to submit mail at this time");
	if (rp_isbad (mm_winit ((char *) 0, subflags, from)))
		syserr("Problem with submit message initialization");
	if (!extract) {
		if (rp_isbad (retval = mm_rrply (&thereply, &len)))
			syserr("Initialization reply failure");
		switch (rp_gbval (thereply.rp_val)) {
		case RP_BNO:
		case RP_BTNO:
			syserr("Initialization failure: %s", thereply.rp_line);
		}
	}

	if (*av == NULL && !extract) {
		syserr("Usage: /usr/sbin/sendmail [flags] addr...");
	}
	if (!extract) {
		while (*av)
			send_address(*av++);
		if (rp_isbad (mm_waend ()))
			syserr("Problem with address list.");
	}
	if (verify) {
		mm_end(NOTOK);
		exit(badaddrs ? 1 : 0);
	}

	doheader();

	exit(dobody());
}

send_address (addr)
char    *addr;
{
	struct rp_bufstruct thereply;
	int     retval;
	int     len;

	if (watch) {
		printf ("%s:  ", addr);
		fflush (stdout);
	}

	if (rp_isbad (retval = mm_wadr ((char *)0, addr)) ||
	    rp_isbad (retval = mm_rrply (&thereply, &len)))
		syserr("Problem in send_address [%s].", rp_valstr (retval));
	switch (rp_gval (thereply.rp_val)) {
	case RP_AOK: 
		if(watch) printf ("address ok\n");
		break;

	case RP_NO: 
		if(watch) printf ("not deliverable; unknown problem\n");
		badaddrs = TRUE;
		break;

	case RP_USER: 
		if(watch) printf ("not deliverable; unknown address.\n");
		badaddrs = TRUE;
		break;

	case RP_NDEL: 
		if(watch) printf ("not deliverable; permanent error.\n");
		badaddrs = TRUE;
		break;

	case RP_AGN: 
		if(watch) printf ("failed, this attempt; try later\n");
		badaddrs = TRUE;
		break;

	case RP_NOOP: 
		if(watch) printf ("not attempted, this time; perhaps try later.\n");
		badaddrs = TRUE;
		break;

	default: 
		syserr("Unexpected address response:  %s", thereply.rp_line);
	}
	fflush (stdout);
}

doheader()
{
	int	gotfrom, gotsender, gotdate;
	char	line[LINESIZE];

	gotfrom = gotsender = gotdate = 0;
	while (fgets (line, LINESIZE, stdin) != NULL) {
		if (line[0] == '\n')
			break;
		if (prefix ("Date:", line))
			gotdate++;
		if (prefix ("From:", line)) {
			gotfrom++;
			if (rewritefrom) {
				dofrom();
				continue;
			}
		}
		if (prefix ("Sender:", line))
			gotsender++;
		mm_wtxt (line, strlen (line));
	}
	if (!gotdate) {
		strncpy (line, "Date: ", sizeof(line));
		cnvtdate (TIMREG, line+6);
		strcat (line, "\n");
		mm_wtxt (line, strlen(line));
	}
	if (!gotfrom)
		dofrom();
	if (!gotsender) {
		snprintf(line, sizeof(line), "Sender: %s\n", from);
		mm_wtxt (line, strlen(line));
	}
	mm_wtxt("\n", 1);
}

dofrom()
{
	char	line[128];

	if (isstr(FullName))
		snprintf(line, sizeof(line), "From: %s <%s>\n", FullName, from);
	else
		snprintf(line, sizeof(line), "From: %s\n", from);
	mm_wtxt(line, strlen(line));
}

dobody()
{
	struct rp_bufstruct thereply;
	char    buffer[BUFSIZ];
	register int    i;
	int     len;

	while (!feof (stdin) && !ferror (stdin) &&
	    (i = fread (buffer, sizeof (char), sizeof (buffer), stdin)) > 0)
		if (rp_isbad (i = mm_wtxt (buffer, i)))
			syserr("Problem writing body");

	if (ferror (stdin))
		syserr("Problem reading body");

	if (rp_isbad (mm_wtend ()) || rp_isbad (mm_rrply (&thereply, &len)))
		syserr("problem ending submission");

	if (rp_isbad (thereply.rp_val))
		syserr("%s", thereply.rp_line);

	return(0);	/* eventually the program exit value */
}

smtp()
{
	char	*smtpd = dupfpath(SMTPSRVR, chndfldir);
	static char fqdn[512];
 
	fqdn[511]='\0';
	strncpy(fqdn,locname,511);
	strncat(fqdn,".",511-strlen(fqdn));
	strncat(fqdn,locdomain,511-strlen(fqdn));

#ifdef HAVE_SETREUID
	setreuid(geteuid(),geteuid());
#else /* HAVE_SETREUID */
	setuid(geteuid());	/* Must become "mmdf" for real */
#endif /* HAVE_SETREUID */
	execl (smtpd, "sendmail-smtp", from, fqdn, "local", (char *)0);
	perror(smtpd);
	exit(9);
}

/*VARARGS1*/
syserr(fmt, a, b)
char	*fmt, *a, *b;
{
	fprintf(stderr, fmt, a, b);
	fputc('\n', stderr);
	exit(9);
}

RETSIGTYPE die(sig)
int sig;
{
	mm_end(NOTOK);
	fprintf(stderr, "sendmail: dying from signal %d\n", sig);
	exit(99);
}
